### ISSUE TITLE
refactoring: DataLoader returns a DataLoaderWrapper?

### ISSUE THREAD

**Issue by jeff-regier** [YosefLab/scVI#156]
That seems backwards: the wrapper usually knows about class it wraps, not the other way around. Is there a way to just combine DataLoader and DataLoaderWrapper into one class? I don't think we ever want to use a dataloader without calling the ability to call the `tocuda` method.

https://github.com/YosefLab/scVI/blob/c08d115435499749d599f66793d4278be232ca26/scvi/dataset/data_loaders.py#L77-L88

**Comment by Edouard360** [comment]
`DataLoaders` handles multiple data_loaders of class `DataLoaderWrapper`.
`DataLoaderWrapper` wraps `DataLoader` of `torch.utils.data`. 

`DataLoaderWrapper` is not wrapping `DataLoaders` so it's not backward, is it ? I just want to easily create data_loaders with the same structure (ie, batch_size, pin_memory, ect...), that's why I created DataLoaders class. Maybe I should rename that function to something more explicit like `new_data_loader` ?

**Comment by jeff-regier** [comment]
OK, I see....that's pretty tricky. What I'm about to propose is kind of a big change conceptually, but I think it might clean everything up, and that it wouldn't actually take too much time.

Rather than having `VariationalInference` take a `Dataset` object as an argument, it takes one `DataLoader` (i.e of `torch.util.data`). It no longer inherits from `Inference`, and we rename `Inference` to `Trainer`. `VariationalInference` is renamed to `Posterior`. Methods inside `Posterior` no longer need `name` as an argument because they always get applied to one epoch of the data loader, e.g.

https://github.com/YosefLab/scVI/blob/1e07545d29ab804062deeadfb43476ae200faf4f/scvi/inference/variational_inference.py#L363

Each `Trainer` has multiple `Posterior` objects, as well as a train method. One `Posterior` object for the training set, another for the testing set, and yet another for the whole dataset if necessary. Or, in the case of `SmFishTrainer` (for the smFISH data) if has different `Posterior` objects for the different dataset. `SemiSupervisedTrainer` and `SmFishTrainer` both inherit from `Trainer`. Maybe there's also `UnsupervisedTrainer` class (for the standard VAE) and `Trainer` is a common base class for all three.

Then all this `DataLoaders` (plural) stuff goes away entirely -- absorbed by the `Trainer` classes (including the subclasses).

Do you think that could work? I know it might seem like overkill but I also have in mind the upcoming work on spatial DE.

**Comment by jeff-regier** [comment]
So a session of scVI with the VAE model might be like
```
vae = VAE(...)
gene_dataset = CortexDataset()
trainer = Trainer(vae, gene_dataset)
trainer.train()
trainer.test_set.ll(verbose=True)
trainer.test_set.entropy_batch_mixing(verbose=True)
```

Here `trainer.test_set` is an instance of `Posterior`. 
There could be `Posterior` objects like `trainer.train_set` and `trainer.full_dataset` too.

**Comment by Edouard360** [comment]
I see, I think it would work pretty ok. From what I understand, the two big changes are:

- it doesn't change much the logic of the data_loaders, but moves them inside `Trainer`, and first refers to a `set` (pytorch data_loaders), before calling a "task", or "metric", rather than the other way around. I think this seems more easy to understand for the user, indeed ! ("what is sequential and how do I get it ?"). Yet keeping `DataLoaderWrapper` ?

- It separates the training logic, from the task available through the posterior object (`. entropy_batch_mixing `, ect).

**Comment by jeff-regier** [comment]
Yes, that's it. I guess `DataLoaderWrapper` should stick around, but maybe we could call it something more descriptive, like `ToCudaDataLoader`. Also, it'd be good to move what's now called `DataLoaderWrapper` to the same file that contains the `Trainer` classes.

**Comment by Edouard360** [comment]
The `.imputation` task needs to handle multiple data_loaders at once. I wonder if it should have it own `ImputationTrainer` subclass. If we were to include it simply as a `.imputation` task, it would need to be a method of `Trainer`, since a `Posterior` object only refers to one data_loader.
It would have two posteriors one corresponding to the corrupted data, and another to the normal data.

**Comment by jeff-regier** [comment]
Just imputing based on the current model shouldn't take two data loaders. It just outputs the median based on the current model.

Are you talking about training the model do a better job at imputation, by training it with corrupted data? That would seem to require a different Trainer, e.g., `ImputationTrainer`.
